Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-            elementToAdjust->updateEffectiveDirectionalityOfDirAuto();
-            return;
+    if (RefPtr shadowRoot = element.shadowRoot()) {
+        for (Ref child : childrenOfType<Element>(*shadowRoot)) {
+            if (child.ptr() == exclude)
+                continue;
+
+            updateElementAndShadowEffectiveTextDirection(child, direction);
+            child->updateEffectiveTextDirectionOfDescendants(direction);
         }
     }
+
+    if (element.renderer() && element.renderer()->style().direction() != effectiveDirection)
+        element.invalidateStyleForSubtree();
 }
 
-void Element::updateTextDirectionalityAfterInputTypeChange()
+static void updateElementTextDirection(Element& element, TextDirectionState textDirectionState, Element* exclude = nullptr)
 {
-    dirAttributeChanged(attributeWithoutSynchronization(dirAttr));
+    setElementHasDirAutoFlag(element, textDirectionState);
+
+    auto direction = computeElementTextDirection(element, textDirectionState);
+
+    updateElementAndShadowEffectiveTextDirection(element, direction, exclude);
+    element.updateEffectiveTextDirectionOfDescendants(direction, exclude);
+
+    if (RefPtr parent = element.parentOrShadowHostElement()) {
+        if (parent->selfOrPrecedingNodesAffectDirAuto())
+            parent->updateEffectiveTextDirectionOfAncestor(&element);
+    }
 }
 
-void Element::updateEffectiveDirectionalityOfDirAuto()
+bool Element::hasValidTextDirectionState() const
 {
-    auto result = computeDirectionalityFromText();
-    setHasDirAutoFlagRecursively(this, true, result.strongDirectionalityNode.get());
-    updateEffectiveDirectionality(result.direction);
-    if (renderer() && renderer()->style().direction() != result.direction)
-        invalidateStyleForSubtree();
+    return elementHasValidTextDirectionState(*this);
 }
 
-auto Element::computeDirectionalityFromText() const -> TextDirectionWithStrongDirectionalityNode
+bool Element::hasAutoTextDirectionState() const
 {
-    if (RefPtr textControl = dynamicDowncast<HTMLTextFormControlElement>(const_cast<Element*>(this))) {
-        if (textControl->dirAutoUsesValue()) {
-            auto direction = textControl->value().defaultWritingDirection();
-            if (!direction)
-                return { TextDirection::LTR, nullptr };
-            return { *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL, WTFMove(textControl) };
-        }
-    }
+    return elementHasAutoTextDirectionState(*this);
+}
 
-    RefPtr node = firstChild();
-    while (node) {
-        // Skip bdi, script, style and text form controls.
-        auto* element = dynamicDowncast<Element>(*node);
-        if (node->hasTagName(bdiTag) || node->hasTagName(scriptTag) || node->hasTagName(styleTag)
-            || (element && element->isTextField())) {
-            node = NodeTraversal::nextSkippingChildren(*node, this);
+std::optional<TextDirection> Element::computeAutoTextDirection() const
```

Comment:
This is computing TextDirection when dir=auto is set. We should call it something like `computeTextDirectionForDirAuto` or something.

Reply:
This is not necessarily the case. The specs says: to compute the text direction in the case of dir _undefined_:

> [undefined](https://html.spec.whatwg.org/multipage/dom.html#attr-dir-undefined-state)
> If element is a [bdi](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-bdi-element) element
> Let result be the [auto directionality](https://html.spec.whatwg.org/multipage/dom.html#auto-directionality) of element.
>

Reply:
Okay, in that case, we should probably call it something like `computeTextDirectionForAutoTextDirectionState`.

Reply:
Hm... maybe we can just refer to https://html.spec.whatwg.org/multipage/dom.html#auto-directionality and call this "computeAutoDirectionality" to match the spec language?